%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Nick Waters's super awesome template for biorXiv submissions

% In memory of Henry, Hermione, and Angus.
% May there be many tasty snails in pufferfish heaven
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[10pt]{article}

\usepackage{geometry}
\geometry{marginparwidth=1cm,a4paper,verbose,tmargin=1.5cm,bmargin=1.5cm,lmargin=1.5cm,rmargin=1.5cm,headheight=0cm,headsep=0cm,footskip=0cm}
\usepackage{cite}
\usepackage{hyperref} % make the references clickable
\setlength\parindent{0pt} % set indent to zero
\setlength{\parskip}{1.2em} % give a bit of space between paragraphs

\bibliographystyle{plain} % citation and bib style
\usepackage{grffile} %for underscores in file names
\usepackage[ruled]{algorithm2e} % for typesetting the riboSeed algorithm
\setlength{\algotitleheightrule}{0pt}
\usepackage{lineno, color} % controls line numbering
\modulolinenumbers[5]  % number every 5th line
% format our line numbers, cause no one likes boring numbers
\renewcommand\linenumberfont{\normalfont\tiny\color{blue}}
\usepackage{booktabs}  % less yucky tables
% nice little thing for table spacing (thanks Markus Püschel)
\newcommand{\ra}[1]{\renewcommand{\arraystretch}{#1}}
\usepackage{multirow} % for some pretty tables with merged cells

% my lined-off abstract
\newenvironment{myabstract}{%
\begin{quote} \baselineskip14pt \rule{.89\textwidth}{1.5pt} \vskip .1cm}
{ \vskip -.10cm \rule{.89\textwidth}{1.5pt} \end{quote}}

% supplementary (essentialy just resets numbering)
\newcommand{\beginsupplement}{%
  \setcounter{table}{0}
  \renewcommand{\thetable}{S\arabic{table}}%
  \setcounter{figure}{0}
  \renewcommand{\thefigure}{S\arabic{figure}}%
}
% pretty thought break
\def \thoughtbr {\begin{center}\noindent\rule{.4\textwidth}{0.4pt}  {\raisebox{-.5ex}{$\sim$}}  \rule{.4\textwidth}{0.4pt}\end{center}}

% \usepackage{graphicx}
\graphicspath{ {riboFigs/} }  %  set the path to the figures

\usepackage{titlesec} % for adjusting spacing after the section headers
\titlespacing\subsection{0pt}{12pt plus 4pt minus 2pt}{-11pt plus 0pt minus 0pt}
\titlespacing\subsubsection{0pt}{12pt plus 4pt minus 2pt}{-11pt plus 0pt minus 0pt}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% this hack stolen from Stack overflow is a macro to make a \cmmidrule
% for each column.  I added the ".4" spacing to get this to look prettier
%  if you have 5 columns, instead of \midrule, you could use \cmidrules{5}
\makeatletter
\newtoks\MD@cmidrules
\newcommand{\cmidrules}[1]{%
  \noalign{%
    \global\MD@cmidrules={}%
    \toks@={\cmidrule(l{.3\tabcolsep}r{.3\tabcolsep})}%
    \count@=\z@
    \loop\ifnum\count@<#1\relax
      \advance\count@\@ne
      \edef\MD@temp{\the\toks@{\the\count@-\the\count@}}%
      \global\MD@cmidrules\expandafter{\the\expandafter\MD@cmidrules\MD@temp}%
    \repeat
  }%
  \the\MD@cmidrules
}
\makeatother
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%  This allows us to set the table font size up top
\usepackage{etoolbox}
\BeforeBeginEnvironment{tabular}{\begin{center}\footnotesize}
\AfterEndEnvironment{tabular}{\end{center}}
%%%%

\def \coli#1 {\textit{E.~coli~#1~}} %  #lazy
\def \ttilde {\raisebox{-.5ex}\textasciitilde} % fix for ugly cmodern tilde 20160829

%  format captions
\usepackage[%
    font={small},
    labelfont=bf,
    format=hang,
    format=plain,
    margin=0pt,
    width=0.9\textwidth]{caption}
\usepackage[list=true]{subcaption}  % allow for subfigures with captions
\usepackage{amssymb}  % for checkmark symbols
\usepackage{graphicx}  % trim figures
\usepackage{colortbl}  % adjust table colors
\definecolor{lgray}{gray}{.90}  % line grey
\definecolor{tgray}{gray}{.40}  % text grey
\definecolor{egray}{gray}{.01}  % emphasis grey
\usepackage[yyyymmdd,hhmmss]{datetime} % for timestamp

\title{\fontseries{m}\selectfont riboSeed: leveraging prokaryotic genomic architecture to assemble across ribosomal regions}

\author
{\small Nicholas R Waters,$^{1,2\ast}$ Florence Abram,$^{1}$ Fiona Brennan,$^{1,3}$  Ashleigh Holmes,$^{2}$ and Leighton Pritchard$^{2}$\\
\\
\normalsize{\textit {$^{1}$National University of Ireland, Galway}}\\
\normalsize{\textit {$^{2}$The James Hutton Institute, Dundee, Scotland}}\\
\normalsize{\textit {$^{3}$Teagasc, Johnstown Castle, Wexford}}\\
\\
\footnotesize{$^\ast$To whom correspondence should be addressed; E-mail: nickp60@gmail.com}
}
\date{}

\begin{document}
\baselineskip20pt  % set line spacing; double spacing is 2xfontsize, in this case 11
\maketitle
\vspace*{-1.5cm}
{\begin{center}\footnotesize\date{Compiled: \today ~\currenttime}\end{center}}


\begin{myabstract}
The vast majority of bacterial genome sequencing has been performed using Illumina short reads. Because of the inherent difficulty of resolving repeated regions with short reads alone, only 10\% of sequencing projects have resulted in a closed genome. The most common repeated regions are those coding for ribosomal operons (rDNAs), which can occur in a bacterial genome between 1 and 15 times and are typically used as sequence markers to classify and identify bacteria. Here, we show that the genomic context in which rDNAs occur is conserved across taxa and that, by utilizing the conserved nature of rDNAs across taxa and the uniqueness of their flanking regions, it is possible to improve assembly of these regions relative to \textit{de novo} sequencing. We describe a method which constructs targeted pseudocontigs generated by iteratively assembling reads that map to a reference genome’s rDNAs. These pseudocontigs are then used to more accurately assemble the newly-sequenced chromosome. We show that this method, implemented as riboSeed, correctly bridges across adjacent contigs in bacterial genome assembly and, when used in conjunction with other genome polishing tools, can result in closure of a genome.
\end{myabstract}

Keywords: genome assembly, ribosome, benchmarking, scaffolding, \textit{de fere novo}

\begin{linenumbers}


\section*{Background}

\begin{table}[!b]
  \centering
  \ra{1.3}
  \caption{NCBI Genome Assemblies of Bacteria}
  \label{table:completions}
  \begin{tabular}{lrrrrr}
    \toprule
    Date & Total & Complete & Chromosome & Scaffold & Contig \\
    \cmidrules{6}
    January 4\textsuperscript{th}, 2017 & 85799 & 6255 & 1143 & 39972 & 38429  \\
    May 17\textsuperscript{th}, 2017 & 96849 & 7212 & 1254 & 42839 & 43899\\
    \bottomrule
    \multicolumn{3}{l}{\tiny Source: \url{https://www.ncbi.nlm.nih.gov/genome/browse/}}
  \end{tabular}
\end{table}

Sequencing bacterial genomes has become much more cost effective and convenient, but the number of complete, closed bacterial genomes remains a small fraction of the total number sequenced (Table \ref{table:completions}). The length of short reads is increasing, but even with the advent of new long-read technologies, bacterial assembly remains a major bottleneck\cite{Nagarajan2010,Brouwer2016}. Although draft genomes are often of very high quality and suited for many types of analysis, researchers are forced to choose between working with these draft genomes (and the inherent potential loss of data), or spending time and resources polishing the genome with some combination of \textit{in silico} tools, PCR, optical mapping, re-sequencing, or hybrid sequencing\cite{Nagarajan2010,Utturkar2014}. Many \textit{in silico} genome finishing tools are available, and we summarise several of these in Table \ref{table:tools}.


\begin{table}[!h]
\centering
\ra{1.3}
\caption{Alternative \textit{\textit{in silico}} genome polishing tools}
\label{table:tools}
% \resizebox{\textwidth}{!}{
\begin{tabular}{p{2.6cm}p{11cm}}
  \toprule
  Tool &  Method Summary \\
  \cmidrules{2}
  GapFiller\cite{Boetzer2012} & iteratively utilizes paired-end reads to close contig junctions \\
    \arrayrulecolor{lgray}\hline
  % SOAPdenovo2’s
  GapCloser\cite{Luo2012} & uses paired-end reads to close contig junctions \\
  \hline
  IMAGE \cite{Tsai2010} & iteratively uses local assemblies of reads belonging to  assembly gaps \\
  \hline
  CloG \cite{Yang2011} & use trimmed \textit{de novo} contigs in hybrid assembly followed by a stitching algorithm \\
  \hline
  FGap \cite{Piro2014,Guizelini2016} & uses BLAST to find potential gap closures from alternate assemblies, libraries or references. \\
  \hline
  GFinisher \cite{Guizelini2016} & uses GC-skew to refine assemblies \\
  \hline
  GapFiller \cite{Nadalin2012} &  produces ``long-reads'' from paired-end sequencing data using a local assembler, which can then be used in a \textit{de novo} assembly. \\
  \hline
  CONTIGuator\cite{Galardini2011} & uses contigs from a \textit{de novo} assembly along with one or more reference sequences to generate a contig map and PCR primer sets to validate in the lab. \\
  \hline
  Konnector\cite{Vandervalk2015} & uses paired-end reads to make long reads to be used in a Bloom filter representation of a de Bruijn graph \\
  \hline
  MapRepeat\cite{Mariano2015} & uses a directed scaffolding method to fill in rDNA gaps, but limited to Ion Torrent reads, and affected by inversions between rDNAs \cite{Mariano2016} \\
  \hline
  GRabB\cite{Brankovics2016} & selectively assembles tandem rDNAs and mitochondria\\
    \arrayrulecolor{black}
  \bottomrule
\end{tabular}
% \begin{tabular}{p{1.8cm}p{2.5cm}p{11cm}}
%   \toprule
%   Tool & Reference & Method Summary \\
%   \cmidrules{3}
%   GapFiller & Boetzer, 2012\cite{Boetzer2012} & iteratively utilizes paired-end reads to close contig junctions \\
%     \arrayrulecolor{lgray}\hline
%   % SOAPdenovo2’s
%   GapCloser & Luo, 2012\cite{Luo2012} & uses paired-end reads to close contig junctions \\
%   \hline
%   IMAGE &  Tsai, 2010\cite{Tsai2010} & iteratively uses local assemblies of reads belonging to  assembly gaps \\
%   \hline
%   CloG & Yang, 2011\cite{Yang2011} & use trimmed \textit{de novo} contigs in hybrid assembly followed by a stitching algorithm \\
%   \hline
%   FGap & Piro, 2014\cite{Piro2014}; Guizelini, 2016\cite{Guizelini2016} & uses BLAST to find potential gap closures from alternate assemblies, libraries or references. \\
%   \hline
%   GFinisher & Guizelini, 2016\cite{Guizelini2016} & uses GC-skew to refine assemblies \\
%   \hline
%   GapFiller & Nadalin, 2012\cite{Nadalin2012} & local assembler using a hash-based method to produce ``long-reads'' from paired-end sequencing data, which can then be used in a \textit{de novo} assembly. \\
%   \hline
%   CONTIGuator & Galardini, 2011\cite{Galardini2011} & uses contigs from a \textit{de novo} assembly along with one or more reference sequences to generate a contig map and PCR primer sets to validate in the lab. \\
%   \hline
%   Konnector & Vandervalk, 2015 \cite{Vandervalk2015} & uses paired-end reads to make long reads to be used in a Bloom filter representation of a de Bruijn graph \\
%   \hline
%   MapRepeat & Mariano, 2015\cite{Mariano2015} & uses a directed scaffolding method to fill in rDNA gaps, but limited to Ion Torrent reads, and affected by inversions between rDNA operons \cite{Mariano2016} \\
%   \hline
%   GRabB & Brankovics, 2016\cite{Brankovics2016} & selective assembly tandem rDNA clusters and mitochondria\\
%     \arrayrulecolor{black}
%   \bottomrule
% \end{tabular}
% }
\end{table}



The Illumina entries in NCBI's Sequence Read Archive (SRA) \cite{Kodama2012a} outnumber all other technologies combined by about an order of magnitude (Table \ref{table:searchterms}). Draft assemblies from these datasets have systematic problems common to short read datasets, namely gaps in the sequences due to the difficulty of resolving assemblies of repeated regions\cite{Whiteford2005,Treangen2011}. By improving the ability to resolve assemblies through repeated regions  it may be possible to improve on current assemblies, and therefore obtain additional sequence information from existing short read datasets in the SRA.



The most common repeated regions are those coding for ribosomal RNA operons (rDNAs). Sequencing of the 16S ribosomal region is widely used to identify bacteria and explore microbial community dynamics\cite{Weisburg1991,Clarridge2004,Woese1990,Case2007}, as the region is conserved within taxa, yet retains enough variability to act as a bacterial ``fingerprint'' to separate clades informatively. However, the 16S, 23S, and 5S ribosomal subunit coding regions are often present multiple times in a single prokaryotic genome, and commonly exhibit polymorphism \cite{Coenye2003,Moreno2002,Lukjancenko2010,Vetrovsky2013}. These long, inexactly repeated regions\cite{Alkan2011} are problematic for short-read genome assembly. Other large repeated regions also exist, but none as pervasive as rDNAs, as ribosomes are essential for cell function. As rDNAs are frequently used as a sequence marker for taxonomic classification, resolving their copy number and sequence diversity from short read collections where the assembled genome has collapsed several repeats into a single region could increase the accuracy of community analysis. We present here an \textit{in silico} method, riboSeed, that capitalizes on the genomic conservation of rDNA regions within a taxon to improve resolution of these normally difficult regions and provide a means to benefit from unexploited information in the SRA/ENA short read archives.


riboSeed is most similar in concept to GRabB, the method of Brankovics et. al \cite{Brankovics2016} for assembling mitochondrial and rDNA regions in eukaryotes, as both use targeted assembly. However, GRabB does not make inferences about the number of rDNA clusters present in the genome, or take advantage of their genomic context. In riboSeed, genomic context is resolved by exploiting both rDNA regions and their flanking regions, harnessing unique characteristics of the broader rDNA region within a single genome to improve assembly.


The riboSeed algorithm proceeds from two observations: (1) that although repeated rRNA coding sequences within a single genome are nearly identical, their flanking regions (that is, the neighboring locations within the genome) are distinct, and (2) that the genomic contexts of rDNAs are conserved within a taxonomic grouping. riboSeed uses only reads that map to rDNA regions from a reference genome, and is not affected by chromosomal rearrangements that occur outside the flanking regions immediately adjacent to each rRNA.


Briefly, riboSeed uses rDNA regions from the closest completely sequenced reference genome to generate rDNA cluster-specific ``pseudocontigs'' that are seeded into the raw short reads to generate a final assembly. We refer to this process as \textit{de fere novo} (meaning ``starting from almost nothing'') assembly.


\begin{figure}[!h]
  \centering
  \includegraphics[width=.8\textwidth]{riboSeed_v11}
  \caption{Reads are mapped to a reference genome and those reads that align to rDNA and flanking regions are extracted. A subassembly for each group of reads that maps to an rDNA region is constructed to produce a ``pseudocontig'' for each region. These pseudocontigs are then concatenated together separated by 5kb of Ns as a buffer. Reads are then iteratively mapped to the concatenated pseudocontigs, extracted, and again subassembled to each region. After the final iteration, the pseudocontigs are included with raw reads in a standard \textit{de novo} assembly. The complete process is referred to as \textit{de fere novo} assembly. The subassemblies attempt to bridge proper rDNA regions by ensuring that flanking regions (represented here by colors) remain correctly paired.
  }
  \label{fig:overview}
\end{figure}




\section*{Implementation}
We present riboSeed: a software suite that allows users to easily perform \textit{de fere novo} assembly, given a reference genome sequence and single or paired-end Illumina sequence reads. The code is primarily written in Python3, with several accessory shell and R scripts.


riboSeed relies on a closed reference genome assembly that is sufficiently closely-related to the isolate being assembled, in which rDNA regions are assembled and known to be in the correct context(see Figure \ref{fig:degen}).

riboSeed proceeds in three stages: preprocessing, \textit{de fere novo} assembly, and assessment/visualization.
\subsection*{1. Preprocessing}
\subsubsection*{riboScan.py}
riboScan.py uses Barrnap (\url{https://github.com/tseemann/barrnap}) to annotate rRNAs in the reference genome, and EMBOSS's seqret\cite{Rice2000a} to create GenBank, FASTA, and GFF formatted versions of the reference genome. This preprocessing step unifies the annotation vocabulary for downstream processes.

\subsubsection*{riboSelect.py}
riboSelect.py infers ribosomal operon structure from the genomic location of constituent 16S, 23S and 5S sequences. The number of 16S sequences is identified from the riboScan.py annotation, and Jenks natural breaks algorithm (using the 16S count to set the number of breaks) is then employed to group rRNA annotations into likely operons on the basis of their genomic coordinates. The output identifies individual rDNA clusters and describes their component elements in a plain text file. This output can be easily adjusted by hand before assembly if the clustering does not appear to accurately reflect the arrangement of the operons (for example, based on visualization of the annotations in a genome browser).

\subsection*{2. \textit{De Fere Novo} Assembly}
\subsubsection*{riboSeed.py}
riboSeed.py implements the algorithm described in Figure \ref{fig:algo}. Short reads for the sequenced isolate are mapped to the reference genome using BWA\cite{Li2009}. Reads that map to each annotated rDNA and its flanking regions (default size 1kbp) are extracted into subsets (one per cluster). Each subset of extracted reads (one per cluster) is assembled into a representative pseudocontig with SPAdes\cite{Bankevich2012}, using the reference rDNA regions as a trusted contig. The resulting pseudocontigs are evaluated for inclusion in future mapping/subassembly iterations based on their length (as discussed below), and concatenated into a pseudogenome, in which pseudocontigs are separated by 5kb of Ns as a buffer. This process is repeated in each subsequent iteration, using the pseudogenome as the reference, and the pseudocontig as a trusted contig.


After a specified number of iterations (3 by default), SPAdes is used to assemble all short reads in a hybrid assembly that includes the pseudocontigs from the final iteration as ``trusted contigs'' (or as ``untrusted contigs'' if the mapping quality of reads to that pseudocontig falls below a threshold). As a control, the short reads are also \textit{de novo} assembled without the pseudocontigs.


Although riboSeed uses SPAdes to perform both the subassemblies and the final \textit{de fere novo} assembly, the pseudocontigs could be submitted to any hybrid assembler that accepts short read libraries and contigs. After assembly, the \textit{de fere novo} and \textit{de novo} assemblies are assessed with QUAST\cite{Gurevich2013}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\SetKwProg{Scan}{riboScan}{ }{end}
\SetKwProg{Select}{riboSelect}{ }{end}
\SetKwProg{Seed}{riboSeed}{ }{end}
\begin{figure}[h]
  \centering
  \begin{minipage}{.6\linewidth}
    \begin{algorithm}[H]
      % \KwData{reference, reads}
      %  % initialization\;
      % \Scan{(reference fastas)}{
      % \For{chromosome in reference}{
      % run Barrnap to find rRNAs\;
      % add locus tags to gff\;
      % convert to genbank\;
      % }
      %   combine genbanks\;
      %   return scannedScaffolds\;
      % }
      %   \hrule
      % % \end{algorithm}
      % % \pagebreak
      % % \begin{algorithm}[H]
      %   \Select{(scannedScaffolds)}{
      %   parse coordinated of rRNAs\;
      %   cluster into operons\;
      %   return grouped\_loci\;
      % }
      %   \hrule
      % \end{algorithm}
      % \pagebreak
      % \begin{algorithm}[H]
      %   \Seed{(scannedScaffolds, grouped\_loci, reads, iters)}{
      %   reference = scannedScaffolds\;
      %   clusters = parse grouped\_loci\;
      %   \For{i in iters}{
      %   map reads to reference\;
      %   \For{cluster in clusters}{
      %   filter and extract reads within clusters and flanking\;
      %   subassemble with SPAdes\;
      %   return pseudocontig\;
      % }
      %   assess subassembly\;
      %   \If {success}{
      %   make fauxgenome from pseudocontigs \;
      %   reference = fauxgenome
      %   continue
      % }
      %   \If{last iteration}{
      %   run SPAdes with pseudocontigs\;
      % }
      % }
      % }
      \Seed{(reference, riboSelect\_clusters, reads, iters)}{
        ref = reference\;
        clusters = parse riboSelect\_clusters\;
        \For{i in iters}{
          map reads to ref\;
          \For{cluster in clusters}{
            filter and extract reads within clusters and flanking\;
            subassemble with SPAdes\;
            return pseudocontig\;
          }
          assess subassembly\;
          \If {success}{
            make pseudogenome from pseudocontigs \;
            ref = pseudogenome \;
          }
        }
        run SPAdes with reads and pseudocontigs\;
      }
    \end{algorithm}
  \end{minipage}
  \caption{Pseudocode of riboSeed algorithm}
  \label{fig:algo}
\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 5


\subsection*{3. Assessment and Visualization}
\subsubsection*{riboScore.py}
riboScore.py extracts the regions flanking the rDNAs in the reference and in the assemblies generated by riboSeed. The rDNA flanking regions from an assembly are aligned to rDNA flanking regions reference regions using BLAST, and depending on the scoring of the alignments, calls a junction a correct, incorrect, or ambiguous join based on the criteria outlined below.
\subsubsection*{riboSnag.py}
riboSnag.py is provided as a helper tool to produce useful diagnostics and visualisation concerning rDNA sequence in the reference genome. Using the clustering generated by riboSelect.py, sequences for the clusters can be extracted from the genome, aligned, and Shannon entropy \cite{Schmitt1997} plotted with consensus depth for each position in the alignment.
\subsubsection*{riboSwap.py}
In all cases, we recommend assessing the performance of the riboSeed pipeline visually using Mauve\cite{Darling2004,Darling2011}, Gingr\cite{Treangen2014}, or a similar genome assembly visualizer to compare reference, \textit{de novo}, and \textit{de fere novo} assemblies in addition to riboScore.py. If contigs appear to be incorrectly joined, the offending \textit{de fere novo} contig can be replaced with syntenic contigs from the \textit{de novo} assembly using the riboSwap.py.


\subsubsection*{riboStack.py}
riboStack.py uses bedtools\cite{Quinlan2010} and samtools\cite{Li2009} to compare the depths of coverage in the rDNA regions to randomly sampled regions elsewhere in the reference genome. riboStack.py takes output from riboScan.py, and a BAM file of reads that map to the reference. If the number of riboScan.py-annotated rDNAs matches the number of rDNAs in the sequenced isolate, the coverage depths within the rDNAs will be similar to other locations in the genome. If the coverage of rDNA regions sufficiently exceeds the average coverage elsewhere in the genome, this may indicate that the reference strain has fewer rDNAs than the sequenced isolate. In this case, using an alternative reference genome may produce improved results.

\section*{Results}
\subsection*{Characteristics of rDNA flanking regions}
The ability to use rDNA flanking sequences to uniquely identify and place rDNAs in their genomic context requires the flanking sequences to be distinct within the genome for each region. This is expected to be the case for most, if not all, prokaryotic genomes. We determined that using 1kb flanking widths was sufficient to include differentiating sequence (Figure \ref{fig:blast}). To demonstrate this, rDNA and 1kb flanking regions were extracted from \coli{Sakai} \cite{Hayashi2001} (BA000007.2), in which the rDNAs have been well characterized\cite{Ohnishi2000}. These regions were aligned with MAFFT\cite{Katoh2002}, and their consensus depth and Shannon entropy\cite{Schmitt1997} calculated for each position in the alignment (Figure \ref{fig:entropyA}).


In \coli{MG1655}, the first rDNA is located 363 bases downstream of \textit{gmhB} (locus tag b0200). Homologous rDNA regions were extracted from 25 randomly selected complete \coli{} chromosomes (Table \ref{table:accessions}).  We identified the 20kb region surrounding \textit{gmhB} in each of these genomes, then annotated and extracted the corresponding rDNA and flanking sequences. These sequences were aligned with MAFFT, and the Shannon entropies and consensus depth plotted (Figure \ref{fig:entropyB}).


\begin{figure}[!h]
    \centering
    \begin{subfigure}[b]{.99\textwidth}
      \includegraphics[width=0.9\textwidth]{entropy_plot}
      \caption{rDNAs from \textit{E. coli Sakai}}
      \label{fig:entropyA}
    \end{subfigure}
    \begin{subfigure}[b]{.99\textwidth}
      \includegraphics[width=0.9\textwidth]{entropy_plot_gmbH}
      \caption{Equivalent rDNAs from 25 \textit{E. coli} genomes}
      \label{fig:entropyB}
    \end{subfigure}

    \caption{Consensus coverage depth (grey bars) and Shannon entropy (black points, smoothed entropy as red line) for aligned rDNA regions. For the seven \coli{Sakai} rDNA regions (a), entropy sharply increases moving away from the coding sequences in the rDNA region. In this case flanking regions would be expected to assemble uniquely. By contrast, the rDNA regions occurring closest to homologous \textit{gmbH} genes from 25 \coli{} genomes (b) show greater conservation in their flanking regions. This indicates that flanking regions are more conserved for homologous rDNA than for paralogous operons, and implies that related genomes are useful reference templates for assembling across these regions. (Similar plots for each of the GAGE-B genomes used later for benchmarking can be found in Figure \ref{fig:ent_gage}.)}
    \label{fig:entropy}
\end{figure}

Figure \ref{fig:entropyA} (and Figure \ref{fig:ent_gage}) shows that within a single genome the regions flanking rDNAs are variable between operons. This enables unique placement of reads at the edges of rDNA coding sequences in their genomic context (i.e. there is not likely to be confusion between the placements of rDNA edges within a single genome).


Figure \ref{fig:entropyB} shows that homologous rDNAs, plus their flanking regions, are well-conserved across several related genomes. Assuming that individual rDNAs are monophyletic within a taxonomic group, short reads that can be uniquely placed on a related genome's rDNA as a reference template are also likely able to be uniquely-placed in the appropriate homologous rDNA of the genome to be assembled.


Taken together, these two properties allow for unique placement of reads from homologous rDNA regions in the appropriate genomic context. These ``anchor points'' then reduce the number of branching possibilities in de Bruijn graph assembly for each individual rDNA, and thereby permit complete assembly through the full rDNA region.


\subsection*{Validating Assembly across rDNA regions}

Settings used for analyses in this manuscript are the defaults as of riboSeed version 0.4.09 (except where otherwise noted).


To evaluate the performance of \textit{de fere novo} assembly compared to \textit{de novo} assembly methods, we used Mauve to visualize syntenic regions and contig breaks of the assemblies in relation to the reference genome that was used to generate pseudocontigs. We categorized each rDNA in an assembly as either a success, failure, or misassembly.

An rDNA assembly was classed as correct if two criteria were met: (i) the assembly merged two contigs across a rDNA region such that, based on the reference, the flanking regions of the \textit{de fere novo} assembly were syntenous with those of the reference; and (ii) the assembled contig extends at least 90\% of the flanking length. An assembled cluster was defined as skipped if the ends of one or more contigs aligned within the rDNA or flanking regions (signalling that extension through the rDNA region was not possible). Finally, if two contigs assembled across a rDNA region in a manner that conflicted with the orientation indicated in the reference genome, the rDNA region was deemed to be incorrect.

In all cases, SPAdes was used with the same parameters for both \textit{de fere novo} assembly and \textit{de novo} assembly, apart from the addition of pseudocontigs in the \textit{de fere novo} assembly.

\subsection*{Simulated Reads with Artificial Genome}
To create a small dataset for testing, we extracted 7 distinct rDNA regions from the \textit{E. coli Sakai} genome (BA000007.2), including 5kb upstream and downstream flanking sequence, using the tools riboScan.py, riboSelect.py and riboSnag.py. Those regions were combined to produce a \ttilde100kb artificial test chromosome. pIRS\cite{Hu2012} was used to generate simulated reads (100bp, 300bp inserts, stdev 10, 30-fold coverage, built-in error profile) from this test chromosome. These reads were assembled using riboSeed, with the \coli{MG1655} genome (NC\_000913.3) as a reference.  Because of the random nature of read simulation, this was repeated 8 times.


The \textit{de fere novo} assembly bridged an average of 4 of the 7 rDNA regions in the artificial genome, while the \textit{de novo} assembly method failed to bridge any (Table \ref{fig:simgenome}). To demonstrate that the choice of reference sequence determines the ability to assemble correctly through rDNA regions, we ran riboSeed with the same \coli{} reads using  pseudocontigs derived from the \textit{Klebsiella pneumoniae HS11286} (CP003200.1) genome\cite{Liu2012}. The \textit{de fere novo} assembly with pseudocontigs from \textit{K. pneumoniae} bridged between 1 and 2 rDNAs, but also misassembled  several rDNA gaps (Figure \ref{fig:artificial}).

\begin{figure}[!h]
    \centering
    \hspace*{-2cm}\includegraphics[width=.99\textwidth]{PrettyMauve}
    \caption{Representative Mauve output describing the results of riboSeed assemblies of simulated reads generated by pIRS from the concatenated \textit{E. coli Sakai} artificial genome. From top to bottom: artificial reference chromosome; rRNA annotations (red bars); \textit{de fere novo} assembly (\coli reference), \textit{de novo} assembly (\coli reference), and \textit{de fere novo} assembly (\textit{K. pneumoniae} reference). riboSeed with \coli reference assembles 4 of 7 rDNA regions, but the \textit{de novo} assembly recovers no rDNA regions correctly. riboSeed using a \textit{K. pneumoniae} reference resolves a single rDNA region, but misassembles clusters.
}
    \label{fig:artificial}
\end{figure}

\subsection*{Effect of reference sequence identity on riboSeed performance}
To investigate how riboSeed assembly is affected by choice of reference strain, we implemented a simple mutation model to generate reference sequence variants of the artificial chromosome described above, with a specified level of sequence identity.  A simple model of geometrically-distributed mutations at a desired mutation frequency (Figure \ref{fig:degen_all}) does not address the disparity of conservation between rDNAs and their flanking region, but a second model was also applied wherein substitutions were allowed only to the rDNA flanking regions (Figure \ref{fig:degen_flank}).

\begin{figure}[!h]
  \centering
  \begin{subfigure}[b]{0.99\textwidth}
    \includegraphics[width=0.9\textwidth]{all_degen}
    \caption{Mutations Throughout}
    \label{fig:degen_all}
  \end{subfigure}
  \begin{subfigure}[b]{0.99\textwidth}
    \includegraphics[width=0.9\textwidth]{flanking_degen}
    \caption{Mutations in Flanking Regions Only}
    \label{fig:degen_flank}
  \end{subfigure}

  \caption{Variants of the artificial genome with mutation frequencies between 0 and 0.1 (i.e. 100 substitutions per kbp). Neighbor-joining trees are shown, rooted by the original sequence. Correctly-assembled rDNAs were counted, and the distribution of results shown against the appropriate mutation frequency. Results are shown for models where substitutions are permitted (A) throughout the chromosome, and (B) only the flanking regions, the latter emulating the relative rate of substitution in rDNA and flanking regions. N=8.}
  \label{fig:degen}
\end{figure}

To obtain an estimate of substitution rate for the \textit{E. coli} data used above, Parsnp\cite{Treangen2014} and Gingr\cite{Treangen2014} were used to identify SNPs in the 25 genomes used in the above analysis (Figure \ref{fig:entropy}), with respect to the same region in \coli{Sakai}. An average substitution rate of 0.0062 was observed.

Figure \ref{fig:degen_all} indicates that the more similar the reference sequence is to the genome being assembled, the greater the likelihood of correctly assembling through rDNA regions. When mutating only the flanking regions (Figure \ref{fig:degen_flank}), which more closely resembles the relative mutation frequencies of the rDNA regions, the procedure correctly assembles rDNAs with tolerance to mutation frequencies up to approximately 30 substitutions per kbp.

\subsection*{Simulated reads with \textit{E. coli Sakai} and \textit{K. pneumoniae} Genomes}


To investigate the effect of short read length on riboSeed assembly, pIRS\cite{Hu2012} was used to generate paired-end reads from the complete \coli{MG1655} and \textit{K. pneumoniae NTUH-K2044} genomes, simulating datasets at a range of read lengths appropriate to the most sequencing technology. In all cases, 300bp inserts with 10bp standard deviation and the built-in error profile were used. Coverage was simulated at 20x to emulate low coverage runs and at 50x to emulate coverage close to the optimized values determined by Miyamoto\cite{Miyamoto2014} and Desai\cite{Desai2013}. \textit{De fere novo} assembly was performed with riboSeed using \textit{E. coli Sakai} and \textit{K. pneumoniae HS11286} as references, respectively, and the results were scored with riboScore.py (Figure \ref{fig:simreads}).

\begin{figure}[!h]
  \centering
  \begin{subfigure}[b]{0.45\textwidth}
    \includegraphics[width=0.9\textwidth]{coli}
    \caption{\textit{E. coli}}
    \label{fig:sim_coli}
  \end{subfigure}
  \begin{subfigure}[b]{0.45\textwidth}
    \includegraphics[width=0.9\textwidth]{kleb}
    \caption{\textit{K. pneumoniae}}
    \label{fig:sim_kleb}
  \end{subfigure}

  \caption{Comparison of \textit{de fere novo} assemblies of simulated reads generated by pIRS. In most cases, increasing coverage depth and read length resulted in fewer misassemblies. Assemblies were scored using riboScore.py. N=5.}
  \label{fig:simreads}
\end{figure}



At either 20x or 50x coverage, \textit{de novo} assembly was unable to resolve a single rDNA with any of the simulated read sets. \textit{de fere novo} assembly with riboSeed showed modest improvement to both the \coli{} and \textit{K. pneumoniae} assemblies. Increasing depth of coverage and read length improves rDNA assemblies.

\subsection*{Benchmarking against Hybrid Sequencing and Assembly}

To establish whether riboSeed performs as well with short reads obtained by sequencing a complete prokaryotic chromosome as with simulated reads, we attempted to assemble short reads from a hybrid Illumina/PacBio sequencing project. The hybrid assembly using long reads was able to resolve rDNAs directly, and would provide a benchmark against which to assess riboSeed performance in terms of: (i) bridging sequence correctly across rDNAs, and (ii) assembling rDNA sequence accurately within each cluster.


Sanjar, et al.published the genome sequence of \textit{Pseudomonas aeruginosa BAMCPA07-48} (CP015377.1)\cite{Sanjar2016}, assembled from two libraries: ca. 270bp fragmented genomic DNA with 100bp paired-end reads sequenced on an Illumina HiSeq 4000 (SRR3500543), and long reads from PacBio RS II. The authors obtained a closed genome sequence by hybrid assembly. We ran the riboSeed pipeline on only the HiSeq dataset in order to compare \textit{de fere novo} assembly to the hybrid assembly and \textit{de novo} assembly of the same reads, using the related genome \textit{P. aeruginosa ATCC 15692} (NZ\_CP017149.1) as a reference.

\textit{de fere novo} assembly correctly assembled across all 4 rDNA regions, whereas \textit{de novo} assembly failed to assemble any rDNA regions (Table \ref{table:paohybrid}). Thus, we find that the \textit{de fere novo} assembly using short reads performs better than \textit{de novo} assembly using short reads alone, validated against a complete hybrid assembly using the same data.

\begin{table}[!h]
  {\color{tgray}
  \centering

  \caption{Assembly of Hybrid-Sequenced \textit{P. aeruginosa BAMCPA07-48}}
  \label{table:paohybrid}
  \begin{tabular}{p{4.9cm}p{.95cm}p{.75cm}p{.75cm}p{2.43cm}p{.6cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}}
    % \arrayrulecolor{black}
    \toprule
    \multirow{2}{*}{Sequenced Strain Name} & \multirow{2}{*}{Platform}  & \multirow{2}{*}{Length} & \multirow{2}{*}{Depth}  &  \multicolumn{2}{c}{Reference Strain} &  \multicolumn{3}{c}{\textit{de novo}} & \multicolumn{3}{c}{\textit{de fere novo}} \\
                            &   &   &   & Name & rDNAs  &  $\checkmark$ & skip & $\times$ &  $\checkmark$ & skip & $\times$  \\
    \cmidrule{1-1}\cmidrule(l){2-2}\cmidrule(l){3-3}\cmidrule(l){4-4}\cmidrule(l){5-6}\cmidrule(l){7-9}\cmidrule(l){10-12}
    \textit{P. aeruginosa BAMCPA07-48} & HiSeq & 100 & 200x & ATCC 15692  & 4 & 1 & 3 & 0 & 4 & 0 & 0 \\
    \bottomrule
    \begin{minipage}[t]{.5\textwidth}
      {\tiny
        $\checkmark$ = correct assembly \\ skip = skipped assembly \\ $\times$ = incorrect assembly
      }
    \end{minipage}
  \end{tabular}
  }
\end{table}


\subsection*{Case Study: Closing the assembly of \textit{S. aureus UAMS-1}}
\textit{Staphylococcus aureus UAMS-1} is a well-characterized, USA200, methicillin-sensitive strain isolated from an osteomyelitis patient. The corresponding published genome was sequenced using Illumina MiSeq generating 300bp reads, and the assembly refined with GapFiller as part of the BugBuilder pipeline \cite{Abbott2017}. Currently, the genome assembly is represented by two scaffolds (JTJK00000000), with several repeated regions acknowledged in the annotations\cite{Sassi2015}. As the rDNA regions were not fully characterized in the annotations, we proposed that \textit{de fere novo} assembly could resolve some of the problematic regions.

Using the same reference \textit{S. aureus MRSA252}\cite{Holden2004} (BX571856.1) with riboSeed as was used in the original assembly, \textit{de fere novo} assembly correctly bridged gaps corresponding to two of the five rDNAs in the reference genome (Table \ref{table:U1}). Furthermore, \textit{de fere novo} assembly bridged two contigs that were syntenic with the ends of the scaffolds in the published assembly, indicating that the regions resolved by riboSeed could allow closure of the genome.


\begin{table}[!h]
  {\color{tgray}
  \centering
  \caption{Assembly of the \textit{S .aureus UAMS-1}}
  \label{table:U1}
  \begin{tabular}{p{4.9cm}p{.95cm}p{.75cm}p{.75cm}p{2.43cm}p{.6cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}}
    % \arrayrulecolor{black}
    \toprule
    \multirow{2}{*}{Sequenced Strain Name} & \multirow{2}{*}{Platform}  & \multirow{2}{*}{Length} & \multirow{2}{*}{Depth}  &  \multicolumn{2}{c}{Reference Strain} &  \multicolumn{3}{c}{\textit{de novo}} & \multicolumn{3}{c}{\textit{de fere novo}} \\
                            &   &   &   & Name & rDNAs  &  $\checkmark$ & skip & $\times$ &  $\checkmark$ & skip & $\times$  \\
    \cmidrule{1-1}\cmidrule(l){2-2}\cmidrule(l){3-3}\cmidrule(l){4-4}\cmidrule(l){5-6}\cmidrule(l){7-9}\cmidrule(l){10-12}
    \textit{S.aureus UAMS-1} & MiSeq & 300 & 110x & MRSA252  & 5 &  0 & 5 & 0   & 2 & 3 & 0 \\
    \bottomrule
    \begin{minipage}[t]{.5\textwidth}
      {\tiny
        $\checkmark$ = correct assembly \\ skip = skipped assembly \\ $\times$ = incorrect assembly
      }
    \end{minipage}
  \end{tabular}
}
\end{table}

We modified the BugBuilder pipeline (\url{https://github.com/nickp60/BugBuilder}) used in the published assembly to incorporate pseudocontigs from riboSeed, resulting in a single scaffold of 7 contigs. In this case, riboSeed was able to bring an existing high-quality scaffold to completion.

\subsection*{Benchmarking against GAGE-B Datasets}
We used the Genome Assembly Gold-standard Evaluation for Bacteria (GAGE-B) datasets \cite{Magoc2013} to assess the performance of riboSeed against a set of well-characterized assemblies. These datasets represent a broad range of challenges; low GC content and tandem rDNA repeats prove challenging to the riboSeed procedure. \textit{Mycobacterium abscessus}, having only a single rDNA operon, does not suffer from the issue of rDNA repeats, and was excluded from this analysis.


When a reference used in the GAGE-B study came from the sequenced strain we chose an alternate reference, as using the true reference sequence would provide an unfair advantage to riboSeed. The GAGE-B datasets include both raw and trimmed reads; in all cases, the trimmed reads were used. Results are shown in Table \ref{table:gageb}.

Compared to the \textit{de novo} assembly, \textit{de fere novo} assembly improved the majority of assemblies. In the case of the \textit{S. aureus} and \textit{R. sphaeroides} datasets, particular difficulty was encountered for all of the references tested. In the case of \textit{B. fragilis}, the entropy plot (Figure \ref{fig:ent_frag}) shows that the variability on the 5' end of the operon is much lower than the other strains, likely leading to the misassemblies.

\begin{table}[!h]
  {\color{tgray}
  \centering
  \ra{1.2}
  \caption{Assembly of GAGE-B datasets}
  \label{table:gageb}
  \begin{tabular}{p{4.9cm}p{.95cm}p{.75cm}p{.75cm}p{2.43cm}p{.6cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}>{\color{black}}>{\hfill}p{.4cm}p{.2cm}p{.1cm}}
    % \arrayrulecolor{black}
    \toprule
    \multirow{2}{*}{Sequenced Strain Name} & \multirow{2}{*}{Platform}  & \multirow{2}{*}{Length} & \multirow{2}{*}{Depth}  &  \multicolumn{2}{c}{Reference Strain} &  \multicolumn{3}{c}{\textit{de novo}} & \multicolumn{3}{c}{\textit{de fere novo}} \\
   & & & & Name & rDNAs & $\checkmark$ & skip & $\times$ & $\checkmark$ & skip & $\times$  \\
    \cmidrule{1-1}\cmidrule(l){2-2}\cmidrule(l){3-3}\cmidrule(l){4-4}\cmidrule(l){5-6}\cmidrule(l){7-9}\cmidrule(l){10-12}
    \textit{A. hydrophila SSU}   & HiSeq   & 101   & 250   & ATCC 7966 & 10 & 0 & 10 & 0  & 4 & 6 & 0  \\
    \arrayrulecolor{lgray}\hline
    \textit{B. cereus VD118}    & HiSeq   & 101   & 300  & ATCC 10987  & 12 & 0 & 12 & 0  & 1 & 11 & 0  \\
    \hline
    \textit{B. cereus ATCC 10987}   & MiSeq   & 250   & 100  & NC7401 & 14 & 0 & 14 & 0  & 12 & 2 & 0  \\
    \hline
    \textit{B. fragilis HMW 615}   & HiSeq   & 101   & 250   & 638R & 6 & 0 & 5 & 1  & 0 & 3 & 3  \\
    \hline
    \textit{R. sphaeroides 2.4.1} & HiSeq & 101 & 210 & ATCC 17029  & 4 & 0 & 4 & 0  & 1 & 3 & 0  \\
    \hline
    \textit{R. sphaeroides 2.4.1} & MiSeq & 251 & 100 & ATCC 17029  & 4 & 1 & 2 & 1  & 1 & 2 & 1  \\
    \hline
    \textit{S. aureus M0927} & HiSeq & 101 & 250 & USA300\_TCH1516 & 5 & 0 & 5 & 0  & 3 & 2 & 0  \\
    \hline
    \textit{V. cholerae CO 0132(5)} & HiSeq & 100 & 110 & El Tor str. N16961  & 8 & 0 & 8 & 0  & 5 & 3 & 0  \\
    \hline
    \textit{V. cholerae CO 0132(5)} & MiSeq   & 250   & 100   & El Tor str. N16961 & 8 & 0 & 8 & 0  & 4 & 4 & 0  \\
    \hline
    \textit{X. axonopodis pv. Manihotis UA323} & HiSeq   & 101   & 250   & pv. Citrumelo & 2 & 0 & 1 & 1  & 2 & 0 & 0 \\
    % \arrayrulecolor{black}

    \bottomrule
    \begin{minipage}[t]{.5\textwidth}
      {\tiny
        $\checkmark$ = correct assembly \\ skip = skipped assembly \\ $\times$ = incorrect assembly
      }
    \end{minipage}
  \end{tabular}
  }
\end{table}

\section*{Discussion}
We show that the regions flanking rDNAs from related strains show a high degree of conservation. This homology allows us to infer the location of rDNAs within a newly sequenced isolate, even in absence of the resolution that would be provided by long read sequencing. Comparing the regions flanking rDNAs within a single genome, we observed that with sufficient flanking length, flanking regions show enough variability to differentiate each instance of the rDNAs. Taken together, the cross-taxon homology allows inference of the location (i.e. the flanking regions) of rDNAs, and the variability of these flanking regions within a genome enables identification of reads likely belonging to each cluster.


The similarity between the sequenced isolate and the reference influences the resulting \textit{de fere novo} assembly; distance can be estimated using an alignment-free approach such as the KGCAK database\cite{Wang2015b}. To prevent spurious joining of contigs, if less than 80\% of the reads map to the reference, the resulting pseudocontigs will be treated as ``untrusted'' contigs by SPAdes. However, performance of riboSeed using degenerate artificial genomes shows that although one should use the best reference available for optimal results, the subassembly method is robust against moderate discrepancies between the reference and sequenced isolate's flanking regions.



The method of constructing pseudocontigs implemented by riboSeed relies on having a relevant reference sequence, where the rDNA regions to act as ``bait'', fishing for reads that likely map specifically to that region. Although this has been shown to be an effective way to partition the appropriate reads, perhaps a more robust and supervision-free method would be use a probabilistic representation of equivalent rDNA regions for a particular taxon. By developing a database of hidden Markov Models from each of the rDNAs in a taxon, perhaps the step of choosing an appropriate reference could be circumvented. In the case of datasets where the choice of reference determined riboSeed's effectiveness (see \textit{S. aureus}, table \ref{table:gageb}), a probabilistic approach may improve performance. Figure \ref{fig:entropyA} shows the unique nature of the rDNA flanking regions; however, although Figure \ref{fig:entropyB} shows strong conservation of the 16S region, the 23S and 5S regions show some degree of variation. These areas in particular may benefit from a probabilistic representation.


Several checks are implemented after the subassembly to ensure that the resulting pseudocontig is fit for inclusion. If a subassembly's longest contig is greater than 3x the particular pseudocontig length or shorter than 6kb (a conservative minimum length of a 16S, 23S, and 5S operon), this is taken to be a sign of poor parameter choice so the user is warned, and by default no further seedings will occur to avoid spurious assembly. Such an outcome can be indicative of any of several factors: improper clustering of operons; insufficient or extraneous flanking sequence; sub-optimal mapping; inappropriate choice of k-mer length for subassembly; inappropriate reference; or other issues. If this occurs, we recommend testing the assembly with different k-mers, changing the flanking length, or trying alternative reference genomes. Mapping depth of the rDNA regions is also reported for each iteration; a marked decrease in mapping depth may also be indicative of problems.


Many published genome finishing tools and approaches offer improvements when applied to suitable datasets, but none (including the approach presented in this paper) is able in isolation to resolve all bacterial genome assembly issues. One constraint on the performance of riboSeed is the quality of rRNA annotations in reference strains. Although it is impossible to concretely confirm it is the case \textit{in silico}, we (and others \cite{Mariano2016}) have found several reference genomes of the course of this study that we suspect have collapsed rDNA repeats. We recommend using a tool such as 16Stimator\cite{Perisin2016} or rrnDB\cite{Stoddard2014} to estimate number of 16s (and therefore rDNAs) prior to assembly, or riboStack.py to assess mapping depths after running riboSeed.

As riboSeed relies on de Bruijn graph assembly, the same considerations must be made with both \textit{de novo} and \textit{de fere novo} assembly. Because of this, the results may depend on the choice of read trimming approach, error correction, range of k-mers, or error correction scheme.

The difficulty in determining the accuracy of rDNA counts in reference genome stems from the fact that most genome sequences are released without publishing the reads used to produced the genome. This practice is a major hindrance when attempting to perform coverage-based quality assessment, such as to infer the likelihood of collapsed rDNAs. Data transparency has become the norm for gene expression studies, but that stance has not been universally adopted when publishing whole-genome sequencing results. To ensure the highest quality assemblies, it is imperative the researchers allow the scientific community to scrutinize the raw whole genome sequencing data with the same rigor that would be applied to any other type of high-throughput sequencing project.



\section*{Conclusions}
Demonstration that rDNA flanking regions are conserved across taxa and that flanking regions of sufficient length are distinct within a genome allowed for the development of riboSeed, a \textit{de fere novo} assembly method. riboSeed  utilizes rDNA flanking regions to act as barcodes for repeated rDNAs, allowing the assembler to correctly place and orient the rDNA. \textit{de fere novo} assembly can improve the assembly by bridging across ribosomal regions, and, in cases where rDNA repeats would otherwise result in incomplete scaffolding, can result in closure of a draft genome when used in conjunction with existing polishing tools. Although riboSeed is far from a silver bullet to provide perfect assemblies from short read technology, it shows the utility in using genomic reference data and mixed assembly approaches to overcome algorithmic obstacles. This approach to resolving rDNA repeats unlocks further insights from large public repositories of short read sequencing data, such as SRA, and when used in conjunction with other genome finishing techniques, provides an avenue towards genomes closure.

\end{linenumbers}
\baselineskip13pt

\section*{List of abbreviations}
rDNA: DNA region coding for ribosomal RNA operon; rRNA: ribosomal RNA; SRA: Sequence Read Archive; ENA: European Nucleotide Archive; IG: intergenic, GAGE-B: Genome Assembly Gold-standard Evaluation for Bacteria


\section*{Availability of data and materials}
The riboSeed pipeline and the datasets generated during the current study are available in the riboSeed GitHub repository, \url{https://github.com/nickp60/riboSeed}. The software is released under the MIT licence. Supplementary data can be found in the riboSeed repository under \textbf{\texttt{Waters\_et\_al\_2017}}. The modified BugBuilder pipeline can be found at \url{https://github.com/nickp60/BugBuilder}. Reference strains used for this study can be found in Table \ref{table:strainlist}.

\section*{Competing interests}
The authors declare that they have no competing interests.

\section*{Funding}
The work was funded through a joint project between The James Hutton Institute, Dundee, Scotland, and the National University of Ireland, Galway, Ireland.

\section*{Authors' contributions}
NRW wrote all the bugs.

\section*{Acknowledgements}
We thank Anton Korobeynikov for his helpful tips on optimizing SPAdes. Yoann Augagneur, Shaun Brinsmade, and Mohamed Sassi graciously provided access to the \textit{S. aureus UAMS-1} genome sequencing data.




\pagebreak
\bibliography{riboSeed_refs}
\pagebreak

\beginsupplement
\section*{Supplementary Data}


\subsection*{Making the artificial test genome}
The artificial genome used for testing was constructed using the \texttt{makeToyGenome.sh} script included in the GitHub repository under the \textbf{\texttt{scripts}} directory. Briefly, the 7 rDNA regions from the \coli{Sakai} genome were extracted with 5kb flanking sequence upstream and downstream; these sequences were then concatenated to form a single, \ttilde100kb sequence containing the 7 rDNAs as well as their flanking context.

\subsection*{Archaeal Datasets}
We assessed the effectiveness of riboSeed with assembling archaeal genomes. Most (\ttilde55\%) archaeal genomes have only a single rDNA, and none has been observed to have more than four. As riboSeed requires a sequencing dataset and a reference genome, applicability was limited; of the 104 entries in rrndb with multiple rDNAs, only 7 had multiple entries at the species level. Among those, only 2 had publicly available short read data. We used riboSeed to re-assemble \textit{Methanosarcina barkeri Fusaro DSMZ804} (Ion Torrent PGM, 89bp single-end reads) and \textit{Methanobacterium formicicum st. BRM9} (Illumina HiSeq 2000, 100bp paired-end reads). \textit{Methanobacterium formicicum st. JCM10132} (DRR017790 ) and \textit{Methanosarcina barkeri Fusaro DSMZ804} (SRR2064286) were the only ones that were suitable for riboSeed, meaning that there was publicly available short read data and that there is a related genome at the species level which is complete.


\textit{M. formicicum st. JCM10132} was sequenced on an Ion Torrent PGM, generating 106.5Mbp of single-end data. \textit{M formicicum BRM9} (CP006933.1) was used as a reference. The resulting \textit{de fere novo} assembly resulted in assembly of 1 of 2 rDNA gaps. This represents the first application of riboSeed to Ion Torrent data.


\textit{Methanosarcina barkeri Fusaro DSMZ804} was sequenced using an Illumina HiSeq2000 with 101bp paired-end reads, with an average fragment length of 400bp. We downsampled to use 5\% of the 19.4Gbp dataset. \textit{Methanosarcina barkeri str. Wiesmoor} was used as a reference. The resulting riboSeed assembly showed correct assembly of 3 of 3 rDNAs, while \textit{de novo} assemble failed to resolve any.


Taken together, we show that given appropriate datasets, archaeal datasets can be processed in the same manner used for bacteria.

\thoughtbr

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{table}[!h]
\centering
\caption{Accessions for 25 \textit{E. coli} genomes}
\label{table:accessions}
\begin{tabular}{l}
  \toprule
  GCA\_000021125.1\_ASM2112v1 \\
  GCA\_000023665.1\_ASM2366v1 \\
  GCA\_000026545.1\_ASM2654v1 \\
  GCA\_000262125.1\_ASM26212v1 \\
  GCA\_000273425.1\_Esch\_coli\_MG12655\_V1 \\
  GCA\_000299255.1\_ASM29925v1 \\
  GCA\_000714595.1\_ASM71459v1 \\
  GCA\_000967155.1\_HUSEC2011CHR1 \\
  GCA\_000974405.1\_ASM97440v1 \\
  GCA\_000974465.1\_ASM97446v1 \\
  GCA\_000974575.1\_ASM97457v1 \\
  GCA\_001020945.2\_ASM102094v2 \\
  GCA\_001566675.1\_ASM156667v1 \\
  GCA\_002012245.1\_ASM201224v1 \\
  GCA\_001750845.1\_ASM175084v1 \\
  GCA\_001886755.1\_ASM188675v1 \\
  GCA\_001901145.1\_ASM190114v1 \\
  GCA\_002012145.1\_ASM201214v1 \\
  GCA\_900096815.1\_Ecoli\_AG100\_Sample2\_M9\_Assembly \\
  GCA\_002116715.1\_ASM211671v1 \\
  GCA\_002118095.1\_ASM211809v1 \\
  GCA\_002125925.1\_ASM212592v1 \\
  GCA\_001612475.1\_ASM161247v1 \\
  GCA\_001651965.1\_ASM165196v1 \\
  GCA\_001721125.1\_ASM172112v1\\
  \bottomrule
  {\tiny   All available at \url{ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/}}

\end{tabular}
\end{table}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[]
  \centering
  \caption{Hits resulting from searching the SRA database for various sequencing technologies as of January, 2017}
  \label{table:searchterms}
  \begin{tabular}{lrr}
    \toprule
    Search term & Hits & Percentage \\
    \midrule
    illumina & 2242225 & 94.27 \\
    pacbio & 21131 & 0.89 \\
    ion & 30560 & 1.28 \\
    roche & 42445 & 1.78 \\
    oxford & 12301 & 0.52 \\
    solid & 29791 & 1.25 \\
    \arrayrulecolor{lgray}\hline
    Total & 2378453 & 100\\
    \arrayrulecolor{black}
    \bottomrule
\end{tabular}
\end{table}


\begin{table}[]
\centering
\caption{Strain names and accessions for reference genomes used in this study}
\label{table:strainlist}
\begin{tabular}{ll}
  \toprule
  Strain Name & Accession \\
  \midrule
  \textit{E. coli MG1655} & NC\_000913.3 \\
  \textit{A. hydrophila ATCC 7966} & NC\_008570.1 \\
  \textit{B. cereus ATCC 10987} & AE017194.1 \\
  \textit{B. cereus NC7401} & NC\_016771.1 \\
  \textit{B. fragilis 638R} & FQ312004.1 \\
  \textit{R. sphaeroides ATCC 17029} & NC\_009049.1, NC\_009050.1 \\
  \textit{S. aureus TCH1516} & NC\_010079.1 \\
  % \textit{S. aureus NCTC 8325} & NC\_007795.1 \\
  % \textit{S. aureus FDA209P} & AP014942.1 \\
  \textit{S. aureus MRSA252} & BX571856.1 \\
  \textit{V. cholerae El Tor str. N16961} & NC\_002505.1, NC\_002506.1 \\
  \textit{X. axonopodis pv. Citrumelo} & CP002914.1 \\
  \textit{P. aeruginosa BAMCPA07-48} & CP015377.1 \\
  \textit{P. aeruginosa ATCC 15692} & NZ\_CP017149.1\\
  \bottomrule

\end{tabular}
\end{table}

\begin{table}[]
  \centering
  \caption{Software Versions}
  \label{table:software}
  \begin{tabular}{ll}
    \toprule
    Tool & Version \\
    \midrule
    Mauve & 2015-02-13 build 0 \\
    BLAST+ & 2.2.28+ \\
    Barrnap & 0.7 \\
    BWA & 0.7.12-r1039 \\
    samtools & 1.3.1 \\
    MAFFT & v7.215 \\
    SPAdes & v3.9.0 \\
    QUAST & 4.1 \\
    bedtools & 2.17.0 \\
    EMBOSS & 6.6.0 \\
    pIRS & 2.0.2\\
    \bottomrule
  \end{tabular}
\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\pagebreak

\begin{figure}[!h]
    \centering
    \hspace*{0cm}\includegraphics[width=.8\textwidth]{sakai_BLAST_results}
    \caption{BLASTn was used to perform \textit{in silico} DNA-DNA hybridization of all rDNA regions from \textit{E. coli Sakai} with variable flanking lengths. The number of hits is a proxy for occurrences in the genome; increasing the flanking length increases the specificity.}
    \label{fig:blast}
  \end{figure}

\begin{figure}[!h]
    \centering
    \hspace*{0cm}\includegraphics[width=.60\textwidth]{simulated_genome.pdf}
    \caption{Assembly of artificial genome. \textit{De fere novo} results in closure of 3-5 rDNAs with the correct reference; only 1-2 rDNAs are correctly assembled using \textit{K. pneumoniae}.  No rDNAs are assembled with \textit{de novo} assembly. Scored with riboScore.py. N=8.}
    \label{fig:simgenome}
\end{figure}


\begin{figure}[h]
  \centering
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_000913.3_entropy_plot}
    \caption{\textit{E. coli MG1655} (NC\_000913.3)}
    \label{fig:ent_coli}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/CP003200.1_entropy_plot}
    \caption{\textit{K. pneumoniae subsp. pneumoniae HS11286} (CP003200.1) }
    \label{fig:ent_pneumo}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NZ_CP017149.1_entropy_plot}
    \caption{\textit{P. aeruginosa strain ATCC 15692} (NZ\_CP017149.1)}
    \label{fig:ent_pao}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_008570.1_entropy_plot}
    \caption{\textit{A. hydrophila ATCC 7966} (NC\_008570.1)}
    \label{fig:ent_aero}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/AE017194.1_entropy_plot}
    \caption{\textit{B. cereus ATCC 10987} (AE017194.1)}
    \label{fig:ent_cereus_atcc}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_016771.1_entropy_plot}
    \caption{\textit{B. cereus NC7401} (NC\_016771.1)}
    \label{fig:ent_cereus_nc}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/FQ312004.1_entropy_plot}
    \caption{\textit{B. fragilis 638R} (FQ312004.1)}
    \label{fig:ent_frag}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_009050.1_entropy_plot}
    \caption{\textit{R. sphaeroides  ATCC 17029} (NC\_009049.1, NC\_009050.1)}
    \label{fig:ent_rhodo}
  \end{subfigure}
\end{figure}
\begin{figure}
  \centering
  \ContinuedFloat
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_010079.1_entropy_plot}
    \caption{\textit{S. aureus TCH1516} (NC\_010079.1)}
    \label{fig:ent_tch}
  \end{subfigure}
  % \begin{subfigure}[b]{.45\textwidth}
  %   \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_007795.1_entropy_plot}
  %   \caption{\textit{S. aureus NCTC 8325} (NC\_007795.1)}
  %   \label{fig:ent_nctc}
  % \end{subfigure}
  % \begin{subfigure}[b]{.45\textwidth}
  %   \includegraphics[width=0.95\textwidth]{gage_entropy_figures/AP014942.1_entropy_plot}
  % \caption{\textit{S. aureus FDA209P} (AP014942.1)}
  % \label{fig:entfda}
  % \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/NC_002506.1_entropy_plot}
    \caption{\textit{V. cholerae El Tor str. N16961} (NC\_002505.1) (NC\_002506.1)}
    \label{fig:ent_vib}
  \end{subfigure}
  \begin{subfigure}[b]{.45\textwidth}
    \includegraphics[width=0.95\textwidth]{gage_entropy_figures/CP002914.1_entropy_plot}
    \caption{\textit{X. axonopodis pv. Citrumelo} (CP002914.1)}
  \end{subfigure}
  \caption{riboScan.py,riboSelect.py, and riboSnag.py were run on all the genomes used as references for \textit{de fere novo} assemblies. Consensus coverage depth (grey bars) and Shannon entropy (black points, smoothed entropy as red line) for aligned rDNA regions.}
  \label{fig:ent_gage}

\end{figure}

\end{document}
